proposal/prefixed-parent-selector.md

```diff
@@ -0,0 +1,97 @@
+# Prefixed Parent Selector: Draft 1
+[Issue](https://github.com/sass/sass/issues/1425)
+
+## Background
+
+The parent selector(&) provides flexibility to create rules according to the
+parent classes and simplify wordy expressions to a single line.
+
+It currently supports the variations below:
+```
+& selector { }
+selector & { }
+&selector { }
+```
+However, one interesting way to use the `&` selector is not supported in sass:
+```
+selector& {}
+```
+
+A specific use case for this selector is the following:
+
+```
+.custom-effect {
+  /* common styles */
+}
+
+a.custom-effect {
+  /* styles for <a> element */
+}
+
+button.custom-effect {
+  /* styles for <button> element */
+}
+```
+
+Using the the prefixed parent selector
+```
+.custom-effect {
+
+  // common styles
+
+  a& {
+    // styles for <a> element
+  }
+
+  button& {
+    // styles for <button> element
+  }
+}
+```
+
+Similar functionality is currently supported, but only by using the following verbose syntax:
+```
+.custom-effect {
+
+  // common styles
+
+  @at-root #{selector-unify(&, "a")} {
+    // styles for <a> element
+  }
+
+  @at-root #{selector-unify(&, "button")} {
+    // styles for <button> element
+  }
+}
+```
+
+Another [example](https://github.com/sass/sass/issues/1425#issuecomment-212077907) is included as a comment in the issue about this feature.
+
+This allows for grouping the parent styling based on the different base selectors, while keeping the styling nicely nested within the rest of the ruleset. It is an elegant solution which is not possible without using a combination of built-in sass functions.
+([Current workaround](https://github.com/sass/sass/issues/1425#issuecomment-404462836))
+
+Note that this syntax is supported by Less and Stylus, and people have run into the lack of support for this syntax when trying to switch over to Sass.
+[Comment #1](https://github.com/sass/sass/issues/1425#issuecomment-55462526)
+[Comment #2](https://github.com/sass/sass/issues/1425#issuecomment-123913176)
+[Comment #3](https://github.com/sass/sass/issues/1425#issuecomment-123916395)
+
+There has been an attempt to create a proposal for it, but it fell short in addressing review comments and was closed soon after.
+[Previous Proposal](https://github.com/sass/sass/pull/2723)
+
+## Semantics
+
+The [Style Rules Semantics](https://github.com/sass/sass/blob/master/spec/style-rules.md#semantics) describes the parent selector behavior as follows:
+
+> If `selector` contains one or more parent selectors, replace them with the current style rule's selector and set `selector` to the result.
+
+###Proposal #1:
+
+The current [Style Rules Semantics](https://github.com/sass/sass/blob/master/spec/style-rules.md#semantics) allows parent selectors anywhere in a `selector` and does not explain why it currently errors for prefixed parent selectors. If the prefixed parent selector is supported, the semantics section for Style Rules would not need to be changed.
```

Comment:
This section doesn't say anything about where a parent selector can be because we think of that as a *syntactic* restriction. Sass won't even parse `foo&`, so by the time a style rule is being evaluated we know any parent selectors are in valid locations.

That means that this proposal needs to update the syntax as well as the semantics. This should be clear enough to know whether constructs like `foo&bar`, `&foo&`, `foo&&` and so on are valid syntax.

The semantics should also be made more explicit, similarly to what you've done in Proposal 2. The spec is [written lazily](https://github.com/sass/sass/tree/master/spec#sass-specificiation), so you'll probably need to do some work to specify the existing behavior as well as the new behavior you're adding.

Reply:
Hi, I'm one of @denizsokullu's coworkers and worked on this proposal with him.  

We looked at https://github.com/sass/sass/blob/master/CONTRIBUTING.md#proposal as well as the accepted proposals at https://github.com/sass/sass/tree/master/accepted and were under the impression that either a Syntax or Semantics section was required, not both.  Are you saying that we must provide both in order for this proposal to be accepted?

Also when you say "update the syntax" do you mean that there is an existing syntax specification somewhere?  We weren't able to find one which is why we were proposing that the existing spec supports this syntax.

Reply:
> We looked at https://github.com/sass/sass/blob/master/CONTRIBUTING.md#proposal as well as the accepted proposals at https://github.com/sass/sass/tree/master/accepted and were under the impression that either a Syntax or Semantics section was required, not both. Are you saying that we must provide both in order for this proposal to be accepted?

It depends on what the proposal is changing:

* For proposals like [Plain CSS `min()` and `max()`](https://github.com/sass/sass/blob/master/accepted/min-max.md), the only thing that's changing is how the parser converts the input text into a syntax tree. Once the syntax tree is constructed, the existing language behavior will process it appropriately. In that case, all that's necessary is a "Syntax" section.

* For proposals like [Configuring Modules Through Imports](https://github.com/sass/sass/blob/master/accepted/import-configuration.md), the existing syntax tree remains the same and only the way it's executed changes. In that case, only a "Semantics" section is necessary.

* But most proposals, especially those that add new language features like [`@content` Arguments](https://github.com/sass/sass/blob/master/accepted/content-args.md) or this one, describe both new ways to parse the input file *and* how to handle the newly-parsed structures. In cases like that, both a "Syntax" and "Semantics" section are necessary, respectively.

> Also when you say "update the syntax" do you mean that there is an existing syntax specification somewhere? We weren't able to find one which is why we were proposing that the existing spec supports this syntax.

There's kind of a *theoretical* existing syntax specification. The spec is [added to lazily](https://github.com/sass/sass/blob/master/spec/spec.md#sass-specificiation), so many existing features aren't fully written out until they're first updated by a proposal. For selector parsing, the current syntax would probably look something like this:

```bnf
CompoundSelector       ::= InitialSimpleSelector
                         | InitialSimpleSelector? TrailingSimpleSelector+
InitialSimpleSelector  ::= <type-selector> | '&' <wq-name>?
TrailingSimpleSelector ::= <subclass-selector> | <pseudo-element-selector>
```

...where the angle-bracket productions are defined by [Selectors Level 4](https://www.w3.org/TR/selectors-4/#grammar).

Reply:
> There's kind of a theoretical existing syntax specification. The spec is added to lazily, so many existing features aren't fully written out until they're first updated by a proposal.

If the syntax section only needs to specify how parent selectors are handled we can add that to our proposal. I've never written anything in the grammar language used by web standards, and only rarely read it, so we may need some help fixing mistakes.

> The semantics should also be made more explicit, similarly to what you've done in Proposal 2. The spec is written lazily, so you'll probably need to do some work to specify the existing behavior as well as the new behavior you're adding.

The current [Style Rules Semantics](https://github.com/sass/sass/blob/master/spec/style-rules.md#semantics) say the following about parent selectors:
> If selector contains one or more parent selectors, replace them with the current style rule's selector and set selector to the result.

Do we need to describe the changes to just this line, or do we need to first expand out the entire style rules semantics section?  And if it is the former, what is missing from the change we suggested in our second proposal?

Keep in mind that like many of the other people who want this feature, we are coming from another CSS compiler that supports prefixed parent selectors.  In our case, the system is [PHPSass](https://github.com/richthegeek/phpsass), which supports the Sass syntax from 6 years ago plus some extra stuff like the prefixed parent selector.  As a result our knowledge of Sass's current syntax and semantics is limited to what we've used and what is already documented.  Plus some of what we think we already know about Sass may be inaccurate because of the extra syntax that PHPSass supported.

Reply:
> If the syntax section only needs to specify how parent selectors are handled we can add that to our proposal. I've never written anything in the grammar language used by web standards, and only rarely read it, so we may need some help fixing mistakes.

I'm always happy to help :smile:.

> Do we need to describe the changes to just this line, or do we need to first expand out the entire style rules semantics section? And if it is the former, what is missing from the change we suggested in our second proposal?

The section should be expanded, at least to some degree. You don't need to fully describe the process of resolving nesting, but the process of converting a compound selector that contains parent selectors into a selector list that does not should be explained.